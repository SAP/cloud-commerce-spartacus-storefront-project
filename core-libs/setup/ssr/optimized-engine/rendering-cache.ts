import { SsrOptimizationOptions } from './ssr-optimization-options';

export interface RenderingEntry {
  html?: any;
  err?: any;
  time?: number;
  rendering?: boolean;
}

export class RenderingCache {
  protected renders: {
    [requestKey: string]: RenderingEntry;
  } = {};

  constructor(private options?: SsrOptimizationOptions) {}

  setAsRendering(key: string) {
    this.renders[key] = { rendering: true };
  }

  isRendering(key: string): boolean {
    return !!this.renders[key]?.rendering;
  }

  store(key: string, err?: Error | null, html?: string) {
    this.renders[key] = { err, html };
    if (this.options?.ttl) {
      this.renders[key].time = Date.now();
    }
  }

  get(key: string): RenderingEntry {
    return this.renders[key];
  }

  clear(key: string) {
    delete this.renders[key];
  }

  isReady(key: string): boolean {
    const isRenderPresent = this.renders[key]?.html || this.renders[key]?.err;
    return isRenderPresent && this.isFresh(key);
  }

  isFresh(key: string): boolean {
    if (!this.options?.ttl) {
      return true;
    }

    return Date.now() - (this.renders[key]?.time ?? 0) < this.options?.ttl;
  }
}
