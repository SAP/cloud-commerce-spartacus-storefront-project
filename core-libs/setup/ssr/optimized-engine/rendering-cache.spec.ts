import { RenderingCache } from './rendering-cache';

describe('RenderingCache', () => {
  let renderingCache: RenderingCache;

  beforeEach(() => {
    renderingCache = new RenderingCache({});
  });

  it('should create engine instance', () => {
    expect(renderingCache).toBeTruthy();
  });
});
