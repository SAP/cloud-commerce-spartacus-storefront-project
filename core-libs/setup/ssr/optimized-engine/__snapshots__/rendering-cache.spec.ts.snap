// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderingCache with cacheSize RenderingCache and cacheStrategyResolver if default cacheStrategyResolver should cache if avoidCachingErrors is false 1`] = `
{
  "err": [Error: err],
  "html": "a",
}
`;
